File: player.py
"""
player.py — Player state: location, inventory, score, flags.

The player is you. You are the player. This is your state.
"""

from __future__ import annotations
from dataclasses import dataclass, field

STARTING_ROOM = "docking_bay_7"
MAX_CARRY_WEIGHT = 20


@dataclass
class Player:
    """Encapsulates all mutable player state."""

    current_room: str = STARTING_ROOM
    inventory: list[str] = field(default_factory=list)      # item IDs
    worn_items: list[str] = field(default_factory=list)      # item IDs being worn
    score: int = 0
    turns: int = 0
    health: int = 100
    max_health: int = 100
    flags: dict[str, bool] = field(default_factory=dict)     # story/puzzle flags
    npc_states: dict[str, dict] = field(default_factory=dict)  # per-NPC state
    credits: int = 100

    # ------------------------------------------------------------------
    # Inventory management
    # ------------------------------------------------------------------

    def has_item(self, item_id: str) -> bool:
        return item_id in self.inventory

    def is_wearing(self, item_id: str) -> bool:
        return item_id in self.worn_items

    def add_item(self, item_id: str, weight: int = 0) -> bool:
        """Add item to inventory. Returns False if over weight limit."""
        if self.current_weight() + weight > MAX_CARRY_WEIGHT:
            return False
        self.inventory.append(item_id)
        return True

    def remove_item(self, item_id: str) -> bool:
        """Remove item from inventory. Returns False if not carried."""
        if item_id in self.inventory:
            self.inventory.remove(item_id)
            if item_id in self.worn_items:
                self.worn_items.remove(item_id)
            return True
        return False

    def wear_item(self, item_id: str) -> bool:
        """Mark item as worn. Must already be in inventory."""
        if item_id not in self.inventory:
            return False
        if item_id not in self.worn_items:
            self.worn_items.append(item_id)
        return True

    def current_weight(self) -> int:
        """Return total weight of inventory (placeholder; items track own weight)."""
        return len(self.inventory)  # simplified: each item counts as 1 unit

    # ------------------------------------------------------------------
    # Score and turn tracking
    # ------------------------------------------------------------------

    def add_score(self, points: int) -> None:
        self.score = max(0, self.score + points)

    def tick(self) -> None:
        """Advance one turn."""
        self.turns += 1

    # ------------------------------------------------------------------
    # Flags
    # ------------------------------------------------------------------

    def set_flag(self, flag: str, value: bool = True) -> None:
        self.flags[flag] = value

    def has_flag(self, flag: str) -> bool:
        return self.flags.get(flag, False)

    # ------------------------------------------------------------------
    # Health
    # ------------------------------------------------------------------

    def heal(self, amount: int) -> int:
        """Heal player; returns actual amount healed."""
        old = self.health
        self.health = min(self.max_health, self.health + amount)
        return self.health - old

    def hurt(self, amount: int) -> bool:
        """Damage player; returns True if dead."""
        self.health = max(0, self.health - amount)
        return self.health <= 0

    def is_alive(self) -> bool:
        return self.health > 0

    # ------------------------------------------------------------------
    # NPC state management
    # ------------------------------------------------------------------

    def get_npc_state(self, npc_id: str) -> dict:
        if npc_id not in self.npc_states:
            self.npc_states[npc_id] = {}
        return self.npc_states[npc_id]

    def set_npc_flag(self, npc_id: str, key: str, value: object = True) -> None:
        self.get_npc_state(npc_id)[key] = value

    def get_npc_flag(self, npc_id: str, key: str, default: object = False) -> object:
        return self.get_npc_state(npc_id).get(key, default)

    # ------------------------------------------------------------------
    # Serialization
    # ------------------------------------------------------------------

    def to_dict(self) -> dict:
        return {
            "current_room": self.current_room,
            "inventory": self.inventory,
            "worn_items": self.worn_items,
            "score": self.score,
            "turns": self.turns,
            "health": self.health,
            "max_health": self.max_health,
            "flags": self.flags,
            "npc_states": self.npc_states,
            "credits": self.credits,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Player":
        p = cls()
        p.current_room = data.get("current_room", STARTING_ROOM)
        p.inventory = data.get("inventory", [])
        p.worn_items = data.get("worn_items", [])
        p.score = data.get("score", 0)
        p.turns = data.get("turns", 0)
        p.health = data.get("health", 100)
        p.max_health = data.get("max_health", 100)
        p.flags = data.get("flags", {})
        p.npc_states = data.get("npc_states", {})
        p.credits = data.get("credits", 0)
        return p

File: src/lib/engine/player.js
/**
 * player.js — Player state (port of player.py).
 */

const STARTING_ROOM       = 'docking_bay_7';
const MAX_CARRY_WEIGHT    = 20;

export class Player {
  constructor() {
    this.currentRoom  = STARTING_ROOM;
    this.inventory    = [];   // item IDs
    this.wornItems    = [];   // item IDs being worn
    this.score        = 0;
    this.turns        = 0;
    this.health       = 100;
    this.maxHealth    = 100;
    this.flags        = {};
    this.npcStates    = {};
    this.credits      = 100;
  }

  // ── Inventory ──────────────────────────────────────────────
  hasItem(id)     { return this.inventory.includes(id); }
  isWearing(id)   { return this.wornItems.includes(id); }

  addItem(id, weight = 1) {
    if (this.currentWeight() + weight > MAX_CARRY_WEIGHT) return false;
    this.inventory.push(id);
    return true;
  }

  removeItem(id) {
    if (!this.hasItem(id)) return false;
    this.inventory = this.inventory.filter(i => i !== id);
    this.wornItems = this.wornItems.filter(i => i !== id);
    return true;
  }

  wearItem(id) {
    if (!this.hasItem(id)) return false;
    if (!this.isWearing(id)) this.wornItems.push(id);
    return true;
  }

  currentWeight() { return this.inventory.length; }

  // ── Score / turns ──────────────────────────────────────────
  addScore(pts) { this.score = Math.max(0, this.score + pts); }
  tick()        { this.turns++; }

  // ── Flags ──────────────────────────────────────────────────
  setFlag(flag, value = true) { this.flags[flag] = value; }
  hasFlag(flag)               { return !!this.flags[flag]; }

  // ── Health ─────────────────────────────────────────────────
  heal(amount) {
    const old = this.health;
    this.health = Math.min(this.maxHealth, this.health + amount);
    return this.health - old;
  }
  hurt(amount) {
    this.health = Math.max(0, this.health - amount);
    return this.health <= 0;
  }
  isAlive() { return this.health > 0; }

  // ── NPC state ──────────────────────────────────────────────
  getNpcState(id)                      { if (!this.npcStates[id]) this.npcStates[id] = {}; return this.npcStates[id]; }
  setNpcFlag(id, key, value = true)    { this.getNpcState(id)[key] = value; }
  getNpcFlag(id, key, def = false)     { return this.getNpcState(id)[key] ?? def; }

  // ── Serialization ──────────────────────────────────────────
  toDict() {
    return {
      current_room: this.currentRoom, inventory: this.inventory,
      worn_items: this.wornItems, score: this.score, turns: this.turns,
      health: this.health, max_health: this.maxHealth,
      flags: this.flags, npc_states: this.npcStates, credits: this.credits,
    };
  }

  static fromDict(d) {
    const p = new Player();
    p.currentRoom = d.current_room ?? STARTING_ROOM;
    p.inventory   = d.inventory   ?? [];
    p.wornItems   = d.worn_items  ?? [];
    p.score       = d.score       ?? 0;
    p.turns       = d.turns       ?? 0;
    p.health      = d.health      ?? 100;
    p.maxHealth   = d.max_health  ?? 100;
    p.flags       = d.flags       ?? {};
    p.npcStates   = d.npc_states  ?? {};
    p.credits     = d.credits     ?? 0;
    return p;
  }
}
